File: 368-largest-divisible-subset/368-largest-divisible-subset.java
class Solution {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        int n = nums.length;
        int[] dp = new int[n];
        int[] hash = new int[n];
        for(int i = 0; i < n; ++i) {
            dp[i] = 1;
            hash[i] = i;
        }
        Arrays.sort(nums);
        int maxLIS = 1, lastIdx = 0;
        
        for(int idx = 1; idx < n; ++idx) {
            for(int prev = 0; prev < idx; ++prev) {
                if(nums[idx] % nums[prev] == 0 && dp[idx] < 1 + dp[prev]) {
                    dp[idx] = 1 + dp[prev];
                    hash[idx] = prev;
                }
            }
            if(dp[idx] > maxLIS) {
                maxLIS = dp[idx];
                lastIdx = idx;
            }
        }
        
        List<Integer> lds = new ArrayList<Integer>();
        lds.add(nums[lastIdx]);
        while(lastIdx != hash[lastIdx]) {
            lastIdx = hash[lastIdx];
            lds.add(nums[lastIdx]);
        }
        Collections.reverse(lds);
        return lds;
    }
}

File: 368-largest-divisible-subset/368-largest-divisible-subset.cpp
class Solution {
    // TC: O(n^2) SC: O(n)
public:
    // Using the concept of printing LIS using one pass algorithm and 2 auxilary arrays
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        int n = nums.size();
        // to not check for increasing as well as divisibility for each no. 
        sort(begin(nums), end(nums));
        // dp: used for calculating the LIS, hash: at any index, tells the position of the previous guy to it
        vector<int> dp(n, 1), hash(n);
        iota(begin(hash), end(hash), 0);
        // the count of elements in our LIS, index of the last element in LIS
        int maxLIS = 1, lastIdx = 0;
        for(int i = 1; i < n; ++i) {
            for(int prev = 0; prev < i; ++prev) {
                // check for divisibilty and take into DP
                if(nums[i] % nums[prev] == 0 and dp[i] < 1 + dp[prev]) {
                    // if the count until now exceeds
                    dp[i] = 1 + dp[prev];
                    hash[i] = prev;
                }
            }
            // update the count of elements in our LIS just in case
            if(dp[i] > maxLIS) {
                maxLIS = dp[i];
                lastIdx = i;
            }
        }
        // Longest Divisible Subsequence
        vector<int> lds;
        // obviously
        lds.push_back(nums[lastIdx]);
        // till we reach the last element in LIS
        while(hash[lastIdx] != lastIdx) {
            lastIdx = hash[lastIdx];
            lds.push_back(nums[lastIdx]);
        }
        // reverse to get the required LDS
        reverse(begin(lds), end(lds));
        return lds;
    }
};